tasks/hyrax-metadata-updates/index.js

```diff
@@ -0,0 +1,318 @@
+  // and we should halt.
+  // Either the code is faulty or the provider is trying to ingest granules
+  // into a collection that doesn't exist
+  if (result.length === 0 || isUndefined(result[0].dataset_id)) {
+    throw new RecordDoesNotExist(`Unable to query parent collection entry title using short name ${shortName} and version ${version}`);
+  }
+  return result[0].dataset_id;
+}
+
+/**
+ * generatePath
+ *
+ * @param {Object} config - the config
+ * @param {Object} metadata - the dom
+ * @param {boolean} isUmmG - UMM-G or ECHO10
+ * @throws {Object} invalidArgumentException - if the env is not valid
+ * @returns {string} - the OPeNDAP path
+ */
+async function generatePath(config, metadata, isUmmG) {
+  const providerId = get(config.cmr, 'provider');
+  // Check if providerId is defined
+  if (isUndefined(providerId)) {
+    throw new InvalidArgument('Provider not supplied in configuration. Unable to construct path');
+  }
+  const entryTitle = await getEntryTitle(config, metadata, isUmmG);
+  return `providers/${providerId}/collections/${entryTitle}/granules/${getGranuleUr(metadata, isUmmG)}`;
+}
+
+/**
+ * generateHyraxUrl
+ *
+ * @param {Object} config - the config
+ * @param {Object} metadata - the dom
+ * @param {boolean} isUmmG - UMM-G or ECHO10
+ * @returns {string} - the hyrax url
+ */
+async function generateHyraxUrl(config, metadata, isUmmG) {
+  const path = await generatePath(config, metadata, isUmmG);
+  const url = new URL(`${generateAddress()}/${path}`);
+  return (url.href);
+}
+
+/**
+ * addHyraxUrlToUmmG
+ *
+ * @param {Object} metadata - the metadata dom
+ * @param {string} hyraxUrl - the hyrax url
+ * @returns {string} - the updated metadata containing a Hyrax URL
+ */
+function addHyraxUrlToUmmG(metadata, hyraxUrl) {
+  const metadataCopy = cloneDeep(metadata);
+
+  if (isUndefined(metadataCopy.RelatedUrls)) {
+    metadataCopy.RelatedUrls = [];
+  }
+  const url = {
+    URL: hyraxUrl,
+    Type: 'GET DATA',
+    Subtype: 'OPENDAP DATA',
+    Description: 'OPeNDAP request URL'
+  };
+  metadataCopy.RelatedUrls.push(url);
+
+  return JSON.stringify(metadataCopy, null, 2);
+}
+
+/**
+ * addHyraxUrlToEcho10
+ *
+ * @param {Object} metadata - the metadata dom
+ * @param {string} hyraxUrl - the hyrax url
+ * @returns {string} - the updated metadata containing a Hyrax URL
+ */
+async function addHyraxUrlToEcho10(metadata, hyraxUrl) {
+  const metadataCopy = cloneDeep(metadata);
+
+  // xml2js will model a single child as an element rather than a list so
+  // we have to defend against that by reconstructing OnlineResources as
+  // a list from scratch
+  delete metadataCopy.Granule.OnlineResources;
+  metadataCopy.Granule.OnlineResources = {};
+  metadataCopy.Granule.OnlineResources.OnlineResource = [];
+
+  if (!isUndefined(metadata.Granule.OnlineResources)) {
+    const urls = metadata.Granule.OnlineResources.OnlineResource;
+    metadataCopy.Granule.OnlineResources.OnlineResource.push(urls);
```

Comment:
In this case I think `push` is correct since it will add to the array in `metadataCopy`. `concat` return a new array. I definitely manually tested this scenario and there are unit tests.

Reply:
I believe you and I can see the tests that are passing. I'm still confused though. If `OnlineResource` is an array and you do a `.push`, you'd be adding an array as a new element in the last index of the array, creating a mullti-dimensional array. Do the existing tests assume `OnlineResource` is a single element?